packages/core/typescript/TypeScriptFormattingService.ts

```diff
@@ -0,0 +1,217 @@
+import * as ts from 'typescript';
+import { IFormattingService, IFormatSettings } from '../types';
+import { IFileSystem } from '../types/FileSystem';
+import { Util } from '../util';
+
+const NEW_LINE_PLACEHOLDER = '//I keep the new line';
```

Comment:
Can't quite agree there. The transformer is not a formatter, and it shouldn't be concerned with applying any sorts of format options to the result after parsing the AST. I want to make these two completely separate entities. This could mean that _a)_ the transformer does not print anything on finalize unless a formatter is provided or _b)_ the transformer does print something, but it's not formatted. I chose the latter to keep the finalize method's signature consistent and usable for unit test where we want to just test if the transformer has done its thing without having to explicitly provide a formatter as well.

Regarding the utils I agree that the `getFileSource` and `saveFile` functions are rather useful and do exactly what we want, and I'll be using them in the platform-specific transformers/file updates/whatever you call them and here in the formatter. Though, I should mention that this refactoring effort is not to stop with the file updates refactoring, there's a bigger fish to fry later on when we start implementing our own `vfs`. I have some specific plans for that (which we need to talk about more) and I've been doing this refactoring while keeping in mind what's coming next. Though I agree that some of the decisions made here (like making the formatter use `ts.SourceFIle`) don't make much sense if you don't consider our own `vfs`. So, for the moment I can make it use the regular `App.container` and change that later on if that's fine with you.